### PR TITLE
feature/18868 - Groups for pupil register

### PR DESCRIPTION
**Pull request by pris54**
Populate groups column in pupil's list.

### CODE REVIEW

File: admin/assets/stylesheets/application.scss [hunk 1]
```diff
@@ -416,6 +416,8 @@ table.spacious {
   }
 
   th, td {
+    display: table-cell;
```
Comment (pris54) [inline admin/assets/stylesheets/application.scss]: this line fixes the spacing and aligning problems that appear when a table cell is highlighted (green left border)

File: admin/services/data-access/pupil.data.service.js [hunk 2]
```diff
@@ -171,10 +171,10 @@ pupilDataService.sqlFindPupilsByDfeNumber = async function (dfeNumber, sortDirec
       sortBy = 'lastName'
   }
   const sql = `
-      SELECT 
-        p.*    
+      SELECT p.*, g.group_id 
       FROM ${sqlService.adminSchema}.${table} p 
       INNER JOIN school s ON s.id = p.school_id
+      LEFT JOIN ${sqlService.adminSchema}.[pupilGroup] g ON p.id = g.pupil_id
```
Comment (GeorgeChatzigiannis) [inline admin/services/data-access/pupil.data.service.js]: just one question here: How is the query behaving if you don't have an assigned group?

Comment (GuyHarwood) [reply admin/services/data-access/pupil.data.service.js]: thats what the LEFT JOIN is for

File: admin/controllers/school.js [hunk 3]
```diff
@@ -45,14 +46,22 @@ const getPupils = async (req, res, next) => {
   res.locals.sortOrder = typeof order === 'boolean' ? !order : true
   res.locals.sortClass = order === false ? 'sort up' : 'sort'
   let pupilsFormatted
+  let groupsIndex = []
+
+  try {
+    const groups = await groupService.getGroups(req.user.schoolId)
```
[the remark this answers is not in the document]
Comment (pris54) [reply admin/controllers/school.js]: @GuyHarwood we agreed it returns the id as the dfeNumbe was already available.

File: admin/services/data-access/pupil.data.service.js (hunk 2 above)
Comment (pris54) [reply admin/services/data-access/pupil.data.service.js]: @GeorgeChatzigiannis the left join ensures we still get a successful result, if there are no pupils assigned to groups you'll get a group_id column with null values.

Comment (GeorgeChatzigiannis) [reply admin/services/data-access/pupil.data.service.js]: @pris54 ok just wanted to confirm if it was tested and working as intended basically

File: admin/controllers/school.js [hunk 4]
```diff
@@ -45,14 +45,22 @@ const getPupils = async (req, res, next) => {
...
+    groups.map((obj) => { groupsIndex[obj.id] = obj.name })
```
Comment (jon-shipley) [inline admin/controllers/school.js]: I doubt this line should be in the controller.  The service should do all the work required for the controller.  The service call can make a generic service call then and marshal the result.

Comment (pris54) [reply admin/controllers/school.js]: @jon-shipley getGroups is used in other places, it is only in this case that the mapping is needed. If you wish I can create a new method in the service just to call getGroups and do the mapping? Please confirm.

Comment (jon-shipley) [reply admin/controllers/school.js]: Yes - that sounds like the best approach. 👍

Comment (pris54) [reply admin/controllers/school.js]: @jon-shipley Done.

File: admin/controllers/school.js [hunk 5]
```diff
@@ -44,15 +44,23 @@ const getPupils = async (req, res, next) => {
   const order = JSON.parse(sortOrder)
   res.locals.sortOrder = typeof order === 'boolean' ? !order : true
   res.locals.sortClass = order === false ? 'sort up' : 'sort'
+
   let pupilsFormatted
+  let groupsIndex = []
+
+  try {
```
Comment (GeorgeChatzigiannis) [inline admin/controllers/school.js]: wouldn't it make more sense to place the single line this await call(below) within the try catch block below?
that would also reduce the code a bit.

**Pull request merged**